[Cocoa] Add test coverage for corrupted LocalStorage database

szewai opened the pull request:
#### ce073e6cdcf1092020ccf24df721a5f4dfed0815
<pre>
[Cocoa] Add test coverage for corrupted LocalStorage database
<a href="https://bugs.webkit.org/show_bug.cgi?id=250892">https://bugs.webkit.org/show_bug.cgi?id=250892</a>
rdar://104740551

Reviewed by Chris Dumez.

Add a Cocoa test for 259573@main, which has helped found a null dereference issue in SQLiteStorageArea.

* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
(WebKit::SQLiteStorageArea::prepareDatabase):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LocalStoragePersistence.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/262161@main">https://commits.webkit.org/262161@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ce2764ce3c2d3168a7329ea7c4316c370f8d737

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/706 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/725 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/753 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/923 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/624 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/697 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/774 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/810 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/857 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/713 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/676 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/669 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/886 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/664 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/688 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/637 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/685 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/669 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/659 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/622 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/670 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/179 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/679 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pgorszkowski-igalia commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -165,6 +165,10 @@ bool SQLiteStorageArea::prepareDatabase(ShouldCreateIfNotExists shouldCreateIfNo
+        if (shouldCreateIfNotExists == ShouldCreateIfNotExists::No)
+            return true;
+
+        m_database = makeUnique<WebCore::SQLiteDatabase>();
```
Why you recreate WebCore::SQLiteDatabase?

pgorszkowski-igalia replied:
ah, I found that it is nullptr in handleDatabaseCorruptionIfNeeded in case of corruupted file.

pgorszkowski-igalia commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -165,6 +165,10 @@ bool SQLiteStorageArea::prepareDatabase(ShouldCreateIfNotExists shouldCreateIfNo
     auto openResult  = m_database->open(m_path);
     if (!openResult && handleDatabaseCorruptionIfNeeded(m_database->lastError())) {
         databaseExists = false;
+        if (shouldCreateIfNotExists == ShouldCreateIfNotExists::No)
+            return true;
```
Are you sure that in case of corrupted database file and delete it we don't want to recreate empty database?

szewai replied:
If it's `ShouldCreateIfNotExists::No`, it means the caller does not need to create file or open connection  if database does not exist (for example, the caller wants to perform read operation, and if file is empty then there is nothing to read). 
In the case of corrupted database, after deletion, it's the same as the file does not exist.

cdumez reviewed (approved):
LGTM

szewai committed:
[Cocoa] Add test coverage for corrupted LocalStorage database
https://bugs.webkit.org/show_bug.cgi?id=250892
rdar://104740551

Reviewed by Chris Dumez.

Add a Cocoa test for 259573@main, which has helped found a null dereference issue in SQLiteStorageArea.

* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
(WebKit::SQLiteStorageArea::prepareDatabase):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LocalStoragePersistence.mm:
(TEST):

Canonical link: https://commits.webkit.org/262161@main
